MvvmCross/Core/Core/Navigation/MvxNavigationService.cs

```diff
@@ -171,78 +172,107 @@ private static bool TryGetRoute(string url, out KeyValuePair<Regex, Type> entry)
+        public Task Navigate<TViewModel>() where TViewModel : IMvxViewModel
         {
             var args = new NavigateEventArgs(typeof(TViewModel));
             OnBeforeNavigate(this, args);
             _viewDispatcher.ShowViewModel(new MvxViewModelRequest<TViewModel>());
             OnAfterNavigate(this, args);
+
+            return Task.FromResult(true);
         }
 
-        public async Task<bool> Close(IMvxViewModel viewModel)
+        public Task<bool> Close(IMvxViewModel viewModel)
         {
             var args = new NavigateEventArgs();
             OnBeforeClose(this, args);
             var close = _viewDispatcher.ChangePresentation(new MvxClosePresentationHint(viewModel));
             OnAfterClose(this, args);
-            return close;
+
+            return Task.FromResult(close);
         }
 
-        public Task<TResult> Navigate<TViewModel, TParameter, TResult>(TParameter param)
+        public async Task<TResult> Navigate<TViewModel, TParameter, TResult>(TParameter param)
             where TViewModel : IMvxViewModel<TParameter, TResult>
             where TParameter : class
             where TResult : class
         {
-            throw new NotImplementedException();
+			var args = new NavigateEventArgs(typeof(TViewModel));
+			OnBeforeNavigate(this, args);
+
+			var viewModel = Mvx.IocConstruct<TViewModel>();
+			var request = new MvxViewModelInstanceRequest(viewModel);
+
+			var tcs = new TaskCompletionSource<TResult>();
+			viewModel.SetClose(tcs);
+
+			_viewDispatcher.ShowViewModel(request);
+            await viewModel.Init(param);
+
+			OnAfterNavigate(this, args);
+
+			return await tcs.Task;
         }
 
-        public Task<TResult> Navigate<TViewModel, TResult>()
-            where TViewModel : IMvxViewModelReturn<TResult>
+        public async Task<TResult> Navigate<TViewModel, TResult>()
+            where TViewModel : IMvxViewModelResult<TResult>
             where TResult : class
         {
-            throw new NotImplementedException();
+			var args = new NavigateEventArgs(typeof(TViewModel));
+			OnBeforeNavigate(this, args);
+
+			var viewModel = Mvx.IocConstruct<TViewModel>();
+			var request = new MvxViewModelInstanceRequest(viewModel);
+
+			var tcs = new TaskCompletionSource<TResult>();
+			viewModel.SetClose(tcs);
+
+			_viewDispatcher.ShowViewModel(request);
+			
+			OnAfterNavigate(this, args);
+
+			return await tcs.Task;
         }
 
-        public async Task Navigate<TParameter>(string path, TParameter param)
+        public async Task Navigate<TParameter>(string path, TParameter param) where TParameter : class
         {
             var args = new NavigateEventArgs(path);
             OnBeforeNavigate(this, args);
 
-            var cacheKey = Guid.NewGuid().ToString();
-            NavigationCache.AddValue<TParameter>(cacheKey, param);
-
-            await NavigateRoute(path);
+            var viewModel = await NavigateRoute(path) as IMvxViewModel<TParameter>;
```

Comment:
This either needs to be an explicit cast or we need to handle when `as` returns `null`.